libbeat/beat/event.go

```diff
@@ -77,3 +77,13 @@ func (e *Event) PutValue(key string, v interface{}) (interface{}, error) {
 func (e *Event) Delete(key string) error {
 	return e.Fields.Delete(key)
 }
+
+// Copies the event. NOTE: Private data is not copied
+func (e *Event) Clone() *Event {
```

Comment:
The libbeat/beats package is somewhat public API we (should) try not to change too often. Especially the interfaces on `Event` type are held very small on purpose (methods and helpers around `common.MapStr` have become quite messy).

We've seen a many panics due to unwanted sharing of fields. Instead of asking devs to use Clone (which has been proven to be somewhat error prone) we are thinking about other solutions. For this reason I'd prefer to keep the interface on Event as minimal as possible, so to have less friction here.

The fact that the handling of 'Private' is undefined in the presence of Clone is not very motivating as well. Whatever convention we define, users of the API have to remember it and apply it correctly when writing processors.

One idea is to treat any values/structures added via Put (or some constructor we might define) as readonly internally. We never manipulate potentially shared structs. If values are changed we do some 'magic' (e.g. at worst Clone if that would be required). In order to construct events efficiently one still might prefer common.MapStr (or even go structs in the future), until the moment of freezing the structure has come.

This is also the reason for `common.MapStr` vs. `beat.Event` in the current code base. Heartbeat does a lot of manipulations in order to create events, but the future of Event type/interfaces was not fully clear yet. So I opted for continuing with common.MapStr for most of the operations and defer the transformation `common.MapStr` to `beat.Event` to some point as late as possible.

Note: I'm fully aware of the fact that properly constructing events in heartbeat is extra hard in comparison to any other inputs/collectors/processors we have :(

Reply:
Makes sense. Luckily, I don't even need this method anymore in my new branch.